packages/tpetra/core/src/Tpetra_Details_crsUtils.hpp

```diff
@@ -170,33 +171,40 @@ insert_crs_indices(
     IndexMap&& map,
     std::function<void(size_t const, size_t const, size_t const)> cb)
 {
-  if (new_indices.size() == 0)
+  using offset_type = typename std::decay<decltype (row_ptrs[0])>::type;
+  using ordinal_type = typename std::decay<decltype (cur_indices[0])>::type;
+
+  if (new_indices.size() == 0) {
     return 0;
+  }
 
-  using ordinal = typename InOutIndices::value_type;
-  auto const start = row_ptrs[row];
-  auto end = start + num_assigned;
-  auto num_avail = row_ptrs[row + 1] - end;
+  const offset_type start = row_ptrs[row];
+  offset_type end = start + static_cast<offset_type> (num_assigned);
+  const size_t num_avail = (row_ptrs[row + 1] < end) ? size_t (0) :
+    row_ptrs[row + 1] - end;
+  const size_t num_new_indices = static_cast<size_t> (new_indices.size ());
   size_t num_inserted = 0;
-  for (size_t k = 0; k < new_indices.size(); k++)
-  {
-    auto row_offset = start;
-    ordinal idx = std::forward<IndexMap>(map)(new_indices[k]);
-    for (; row_offset < end; row_offset++)
-      if (idx == cur_indices[row_offset])
+
+  for (size_t k = 0; k < num_new_indices; ++k) {
+    const ordinal_type idx = std::forward<IndexMap>(map)(new_indices[k]);
+    offset_type row_offset = start;
+    for (; row_offset < end; ++row_offset) {
```

Comment:
Any reason to declare the iterated variable outside of the for statement?

Reply:
@lucbv The code uses that variable after the for loop.  It's like `std::find` returning the end iterator.  @tjfulle wrote this code if you have more questions :-)